contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+   v1 backward compatibility is a non-goal. Particularly, scheduling policy
+   config and v1 extenders won't work in this new framework.
+-  Solve all the scheduler v1 limitations, although we would like to ensure
+   that the new framework allows us to address known limitations in the future.
+-  Provide implementation details of plugins and call-back functions, such as
+   all of their arguments and return values.
+
+# DETAILED DESIGN
+
+## Bare bones of scheduling
+
+Pods that are not assigned to any node go to a scheduling queue and sorted by
+order specified by plugins (described [here](#heading=h.dggq0ff44y2y)). The
+scheduling framework picks the head of the queue and starts a **scheduling
+cycle** to schedule the pod. At the end of the cycle scheduler determines
+whether the pod is schedulable or not. If the pod is not schedulable, its status
+is updated and goes back to the scheduling queue. If the pod is schedulable (one
+or more nodes are found that can run the Pod), scoring process is started. The
+scoring process finds the best node to run the Pod.  a bind go routine is
+started to bind the pod.  
+The above process is the same as what Kubernetes scheduler v1 does. Some of the
+essential features of scheduler v1, such as leader election, will also be
+transferred to the scheduling framework.  
+In the rest of this section we describe how various plugins are used to enrich
+this basic workflow. In this section, we focus on in-process plugins.
+Out-of-process plugins are discussed later in the doc.
+
+## Communication and statefulness of plugins
+
+The scheduling framework provides a library that plugins can use to pass
+information to other plugins. This library keeps a map from keys of type string
+to opaque pointers of type interface{}. A write operation takes a key and a
+pointer and stores the opaque pointer in the map with the given key. Other
+plugins can provide the key and receive the opaque pointer. Multiple plugins can
+share the state or communicate via this mechanism.  
+The saved state is preserved only during a single scheduling cycle. At the end
+of a scheduling cycle, this map is destructed. So, plugins cannot keep shared
+state across multiple scheduling cycle. They can, however, update the scheduler
+cache via the provided interface of the cache. The cache interface allows
+limited state preservation across multiple scheduling cycle.  
+It is worth noting that plugins are assumed to be **trusted**. Scheduler does
+not prevent one plugin from accessing or modifying another plugin's state.
+
+## Plugin registration
+
+Plugin registration is done by providing an extension point and a function that
+should be called at that extension point. This step will be something like:
+
+<table>
+<thead>
+<tr>
+<th><p><pre>
+register("pre-filter", plugin.foo)
+</pre></p>
+
+</th>
+</tr>
+</thead>
+<tbody>
+</tbody>
+</table>
+
+The details of the function signature will be provided later.
+
+## Extension points
```

Comment:
Are there any requirements on the plugins regarding time limits?  For example, dynamically provisioning resources could take on the order of minutes.

Reply:
Similarily, current assume cache has a TTL of 30 seconds, which will be too short for dynamic resource provisioning.